The release workflow is not atomic nor idempotent

Right now, the release workflow is not atomic nor idempotent. It's possible that everything goes well for the most part, the npm package gets correctly published, and then the action ends up failing when it tries to create a "new release" in Github.

### Proposed solution:
In order to mitigate this problem, I propose to check if the npm package has been already published before even trying yo publish it. This way, if the previous described scenario happens in real life, we can re-run this action, and it will be able to skip the npm publication and directly jump into the GH release creation, allowing us to fix "unfinished" releases.

I just tried to re-publish one of my own packages, and it seems that npm doesn't like when we try to push the same stuff again:
```
npm ERR! code E403
npm ERR! 403 403 Forbidden - PUT https://registry.npmjs.org/@coderspirit%2fnominal - You cannot publish over the previously published versions: 3.2.1.
```

--- a/src/release.js
+++ b/src/release.js
@@ -76,7 +76,7 @@ module.exports = async function ({ github, context, inputs }) {
     const npmToken = inputs['npm-token']
 
     if (npmToken) {
-      await publishToNpm({ npmToken, opticToken, opticUrl, npmTag })
+      await publishToNpm({ npmToken, opticToken, opticUrl, npmTag, version })
     } else {
       logWarning('missing npm-token')
     }

--- a/src/utils/publishToNpm.js
+++ b/src/utils/publishToNpm.js
@@ -1,7 +1,14 @@
 'use strict'
+
 const { runSpawn } = require('./runSpawn')
 
-async function publishToNpm({ npmToken, opticToken, opticUrl, npmTag }) {
+async function publishToNpm({
+  npmToken,
+  opticToken,
+  opticUrl,
+  npmTag,
+  version,
+}) {
   const run = runSpawn()
 
   await run('npm', [
@@ -10,12 +17,25 @@ async function publishToNpm({ npmToken, opticToken, opticUrl, npmTag }) {
     `//registry.npmjs.org/:_authToken=${npmToken}`,
   ])
 
-  await run('npm', ['pack', '--dry-run'])
-  if (opticToken) {
-    const otp = await run('curl', ['-s', `${opticUrl}${opticToken}`])
-    await run('npm', ['publish', '--otp', otp, '--tag', npmTag])
-  } else {
-    await run('npm', ['publish', '--tag', npmTag])
+  // We need to check if the package was already published. This can happen if
+  // the action was already executed before, but it failed in its last step
+  // (GH release).
+  const pkgInfo = JSON.parse(await run('npm', ['view', '--json']))
+
+  // NPM only looks into the remote registry when we pass an explicit
+  // package name & version, so we don't have to fear that it reads the
+  // info from the "local" package.json file.
+  const allowNpmPublish =
+    '' === (await run('npm', ['view', `${pkgInfo.name}@${version}`]))
+
+  if (allowNpmPublish) {
+    await run('npm', ['pack', '--dry-run'])
+    if (opticToken) {
+      const otp = await run('curl', ['-s', `${opticUrl}${opticToken}`])
+      await run('npm', ['publish', '--otp', otp, '--tag', npmTag])
+    } else {
+      await run('npm', ['publish', '--tag', npmTag])
+    }
   }
 }
 

--- a/test/publishToNpm.test.js
+++ b/test/publishToNpm.test.js
@@ -6,7 +6,18 @@ const proxyquire = require('proxyquire')
 const runSpawnAction = require('../src/utils/runSpawn')
 
 const setup = () => {
-  const runSpawnStub = sinon.stub().returns('otp123')
+  const runSpawnStub = sinon.stub()
+  runSpawnStub
+    .withArgs('curl', [
+      '-s',
+      'https://optic-test.run.app/api/generate/optic-token',
+    ])
+    .returns('otp123')
+  runSpawnStub
+    .withArgs('npm', ['view', '--json'])
+    .returns('{"name":"fakeTestPkg"}')
+  runSpawnStub.withArgs('npm', ['view', 'fakeTestPkg@v5.1.3']).returns('')
+
   const utilStub = sinon.stub(runSpawnAction, 'runSpawn').returns(runSpawnStub)
   const publishToNpmProxy = proxyquire('../src/utils/publishToNpm', {
     './runSpawn': utilStub,
@@ -26,6 +37,7 @@ tap.test('Should publish to npm with optic', async t => {
     opticToken: 'optic-token',
     opticUrl: 'https://optic-test.run.app/api/generate/',
     npmTag: 'latest',
+    version: 'v5.1.3',
   })
 
   sinon.assert.calledWithExactly(runSpawnStub.getCall(0), 'npm', [
@@ -35,19 +47,22 @@ tap.test('Should publish to npm with optic', async t => {
   ])
   t.pass('npm config')
 
-  sinon.assert.calledWithExactly(runSpawnStub.getCall(1), 'npm', [
+  // We skip calls in these checks:
+  // - 1 used to get the package name
+  // - 2 used to check if the package version is already published
+  sinon.assert.calledWithExactly(runSpawnStub.getCall(3), 'npm', [
     'pack',
     '--dry-run',
   ])
   t.pass('npm pack called')
 
-  sinon.assert.calledWithExactly(runSpawnStub.getCall(2), 'curl', [
+  sinon.assert.calledWithExactly(runSpawnStub.getCall(4), 'curl', [
     '-s',
     'https://optic-test.run.app/api/generate/optic-token',
   ])
   t.pass('curl called')
 
-  sinon.assert.calledWithExactly(runSpawnStub.getCall(3), 'npm', [
+  sinon.assert.calledWithExactly(runSpawnStub.getCall(5), 'npm', [
     'publish',
     '--otp',
     'otp123',
@@ -63,6 +78,7 @@ tap.test('Should publish to npm without optic', async () => {
     npmToken: 'a-token',
     opticUrl: 'https://optic-test.run.app/api/generate/',
     npmTag: 'latest',
+    version: 'v5.1.3',
   })
 
   sinon.assert.calledWithExactly(runSpawnStub, 'npm', ['pack', '--dry-run'])
@@ -71,4 +87,71 @@ tap.test('Should publish to npm without optic', async () => {
     '--tag',
     'latest',
   ])
+})
+
+tap.test(
+  'Should skip npm package publication when it was already published',
+  async () => {
+    const { publishToNpmProxy, runSpawnStub } = setup()
+
+    runSpawnStub
+      .withArgs('npm', ['view', 'fakeTestPkg@v5.1.3'])
+      .returns('fake package data that says it was published')
+
+    await publishToNpmProxy.publishToNpm({
+      npmToken: 'a-token',
+      opticUrl: 'https://optic-test.run.app/api/generate/',
+      npmTag: 'latest',
+      version: 'v5.1.3',
+    })
+
+    sinon.assert.neverCalledWith(runSpawnStub, 'npm', [
+      'publish',
+      '--otp',
+      'otp123',
+      '--tag',
+      'latest',
+    ])
+    sinon.assert.neverCalledWith(runSpawnStub, 'npm', [
+      'publish',
+      '--tag',
+      'latest',
+    ])
+  }
+)
+
+tap.test('Should stop action if package info retrieval fails', async t => {
+  t.plan(3)
+  const { publishToNpmProxy, runSpawnStub } = setup()
+
+  runSpawnStub
+    .withArgs('npm', ['view', 'fakeTestPkg@v5.1.3'])
+    .throws(new Error('Network Error'))
+
+  try {
+    await publishToNpmProxy.publishToNpm({
+      npmToken: 'a-token',
+      opticUrl: 'https://optic-test.run.app/api/generate/',
+      npmTag: 'latest',
+      version: 'v5.1.3',
+    })
+  } catch (e) {
+    t.equal(e.message, 'Network Error')
+  }
+
+  sinon.assert.neverCalledWith(runSpawnStub, 'npm', [
+    'publish',
+    '--otp',
+    'otp123',
+    '--tag',
+    'latest',
+  ])
+  t.pass('package is not published with otp code')
+
+  sinon.assert.neverCalledWith(runSpawnStub, 'npm', [
+    'publish',
+    '--tag',
+    'latest',
+  ])
+  t.pass('package is not published without otp code')
 })

--- a/test/release.test.js
+++ b/test/release.test.js
@@ -58,7 +58,15 @@ function setup() {
   const coreStub = sinon.stub(core)
   deleteReleaseStub.resetHistory()
   deleteReleaseStub.resolves()
-  const runSpawnStub = sinon.stub().returns('otp123')
+
+  const runSpawnStub = sinon.stub()
+  runSpawnStub
+    .withArgs('curl', [
+      '-s',
+      'https://optic-test.run.app/api/generate/optic-token',
+    ])
+    .returns('otp123')
+
   const runSpawnProxy = sinon
     .stub(runSpawnAction, 'runSpawn')
     .returns(runSpawnStub)
@@ -159,11 +167,7 @@ tap.test(
       stubs.coreStub.setFailed,
       `Something went wrong while deleting the release. \n Errors: Something went wrong in the release`
     )
-    sinon.assert.neverCalledWith(stubs.runSpawnStub, 'npm', [
-      'publish',
-      '--tag',
-      'latest',
-    ])
+    sinon.assert.notCalled(stubs.publishToNpmStub)
   }
 )
 
@@ -195,18 +199,7 @@ tap.test('Should not publish to npm if there is no npm token', async () => {
     },
   })
 
-  sinon.assert.neverCalledWith(stubs.runSpawnStub, 'npm', [
-    'publish',
-    '--tag',
-    'latest',
-  ])
-  sinon.assert.neverCalledWith(stubs.runSpawnStub, 'npm', [
-    'publish',
-    '--otp',
-    'otp123',
-    '--tag',
-    'latest',
-  ])
+  sinon.assert.notCalled(stubs.publishToNpmStub)
 })
 
 tap.test('Should publish to npm with optic', async () => {
@@ -298,11 +291,7 @@ tap.test("Should NOT use npm if the pr wasn't merged", async () => {
     'sync-semver-tags': 'true',
   }
   await release(data)
-  sinon.assert.neverCalledWith(stubs.runSpawnStub, 'npm', [
-    'publish',
-    '--tag',
-    'latest',
-  ])
+  sinon.assert.notCalled(stubs.publishToNpmStub)
 })
 
 tap.test("Should NOT tag version in git if the pr wasn't merged", async () => {

